### PR TITLE
Accuracy average

### LINKED ISSUE #99: Allow Accuracy to return metric per class
## 🚀 Feature

Implement the `average` argument like in Precision and Recall such that accuracy metric can return the metric per class label.

### Motivation

Sometimes it may be beneficially to look at the accuracy per label, especially when working with very unbalanced datasets

### PR DESCRIPTION
**Pull request by arvindmuralie77**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #99 
Closes #155
I have added the average argument with all other required arguments. For subset_accuracy I have retained some sections of the previous code.
## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@arvindmuralie77 how is it going here, do you think we can squeeze it to 0.3.0?

**Comment by arvindmuralie77** [comment]
Hi @Borda ,

**Comment by Borda** [comment]
@arvindmuralie77 I was just asking what is you estimate to finish this PR :]

**Comment by arvindmuralie77** [comment]
Dear @Borda , I didn't know there was a deadline for 0.3.0. Everything is complete, if default value of `mdmc_average` is set as `global`, then the code is passing on all the previously existing test cases. Now only new test cases need to be created since new parameters have been added to `accuracy`.

**Comment by Borda** [comment]
> I didn't know there was a deadline for 0.3.0

there is no deadline, I just thought that it would be nice to have it there, so just asking how do you feel about it :]

**Comment by arvindmuralie77** [comment]
Hey @Borda , I have added some new test cases for the updated class. I was having some issues understanding the multi-dimension-multilabel class. Apart from that, I think I have covered all other cases. Should I convert this draft to a PR so that you can review it.

File: tests/classification/test_accuracy.py [hunk 1]
```diff
@@ -153,7 +153,7 @@ def test_topk_accuracy(preds, target, exp_result, k, subset_accuracy):
 
     for batch in range(preds.shape[0]):
         topk(preds[batch], target[batch])
-
+    print(topk.compute(), exp_result)
```
Comment (SkafteNicki) [inline tests/classification/test_accuracy.py:156]:
```suggestion

```

File: torchmetrics/classification/accuracy.py [hunk 2]
```diff
@@ -103,15 +152,31 @@ class Accuracy(Metric):
 
     def __init__(
         self,
+        num_classes: Optional[int] = None,
         threshold: float = 0.5,
+        average: str = "micro",
+        mdmc_average: Optional[str] = "global",
+        ignore_index: Optional[int] = None,
```
Comment (SkafteNicki) [inline torchmetrics/classification/accuracy.py:159]: missing docstring for `ignore_index`

File: torchmetrics/classification/accuracy.py [hunk 3]
```diff
@@ -103,15 +152,31 @@ class Accuracy(Metric):
...
+        multiclass: Optional[bool] = None,
         subset_accuracy: bool = False,
         compute_on_step: bool = True,
         dist_sync_on_step: bool = False,
         process_group: Optional[Any] = None,
         dist_sync_fn: Callable = None,
     ):
+        allowed_average = ["micro", "macro", "weighted", "samples", "none", None]
+        if average not in allowed_average:
+            raise ValueError(f"The `average` has to be one of {allowed_average}, got {average}.")
```
Comment (SkafteNicki) [inline torchmetrics/classification/accuracy.py:170]: please add what errors the classes raises to docstring, like:
https://github.com/PyTorchLightning/metrics/blob/b26c20d6f8b89bc03dd4ed8512828e38339d41af/torchmetrics/classification/precision_recall.py#L108-L110

File: torchmetrics/functional/classification/accuracy.py [hunk 4]
```diff
@@ -16,51 +16,122 @@
...
+
+def _mode(
+    preds: Tensor,
+    target: Tensor,
+    threshold: float,
+    top_k: Optional[int],
+    num_classes: Optional[int],
+    multiclass: Optional[bool]
+) -> DataType:
+    mode = _check_classification_inputs(
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/accuracy.py:41]:
```suggestion
    """ returns the mode of the data (binary, multi label, multi class, multi-dim multi class) """
    mode = _check_classification_inputs(
```

**Review by SkafteNicki — COMMENTED** [review]
smaller comments

**Comment by arvindmuralie77** [comment]
@SkafteNicki , I have corrected the mentioned issues.

**Review by SkafteNicki — APPROVED** [review]

File: torchmetrics/classification/accuracy.py [hunk 5]
```diff
@@ -103,15 +163,31 @@ class Accuracy(Metric):
 
     def __init__(
         self,
+        num_classes: Optional[int] = None,
         threshold: float = 0.5,
```
Comment (Borda) [inline torchmetrics/classification/accuracy.py:167]:
```suggestion
        threshold: float = 0.5,
        num_classes: Optional[int] = None,
```
anyway this adding args in middle can be breaking API change for users passing args without key word...

File: tests/classification/test_accuracy.py [hunk 6]
```diff
@@ -200,3 +200,73 @@ def test_wrong_params(top_k, threshold):
...
+
+    # Test functional
+    total_samples = target.shape[0] * target.shape[1]
+
+    preds = preds.view(total_samples, num_classes, -1)
+    target = target.view(total_samples, -1)
+
+    assert (accuracy(
+        preds, target, num_classes=num_classes, average=average, mdmc_average=mdmc_average
+    ) == tensor(exp_result)).all()
```
Comment (Borda) [inline tests/classification/test_accuracy.py:246]:
```suggestion
    acc_score = accuracy(preds, target, num_classes=num_classes, average=average, mdmc_average=mdmc_average) 
    assert (acc_score == tensor(exp_result)).all()
```

File: tests/classification/test_accuracy.py [hunk 7]
```diff
@@ -200,3 +200,73 @@ def test_wrong_params(top_k, threshold):
...
+    assert (acc.compute() == tensor(exp_result)).all()
+
+    # Test functional
+    total_samples = target.shape[0] * target.shape[1]
+
+    preds = preds.view(total_samples, -1)
+    target = target.view(total_samples, -1)
+    assert (accuracy(
+        preds, target, num_classes=num_classes, average=average, mdmc_average=mdmc_average, multiclass=multiclass
+    ) == tensor(exp_result)).all()
```
Comment (Borda) [inline tests/classification/test_accuracy.py:272]:
```suggestion
    acc_score = accuracy(preds, target, num_classes=num_classes, average=average, mdmc_average=mdmc_average, multiclass=multiclass)
    assert (acc_score == tensor(exp_result)).all()
```

File: torchmetrics/classification/accuracy.py [hunk 8]
```diff
@@ -141,18 +220,58 @@ def update(self, preds: Tensor, target: Tensor):
...
-        correct, total = _accuracy_update(
-            preds, target, threshold=self.threshold, top_k=self.top_k, subset_accuracy=self.subset_accuracy
-        )
+        """ returns the mode of the data (binary, multi label, multi class, multi-dim multi class) """
+        mode = _mode(preds, target, self.threshold, self.top_k, self.num_classes, self.multiclass)
+
+        if self.mode is None:
+            self.mode = mode
+        elif self.mode != mode:
+            raise ValueError("You can not use {} inputs with {} inputs.".format(mode, self.mode))
```
Comment (Borda) [inline torchmetrics/classification/accuracy.py:229]: I see here codecov comments that some conditions are not tested, min adding a test for it?
or is it a false alarm?

**Review by Borda — APPROVED** [review]

File: torchmetrics/classification/accuracy.py (hunk 8 above)
Comment (arvindmuralie77) [reply torchmetrics/classification/accuracy.py:229]: Sure, I'll add a test for it.

**Comment by arvindmuralie77** [comment]
Done @Borda :]

File: torchmetrics/classification/accuracy.py [hunk 9]
```diff
@@ -141,18 +220,58 @@ def update(self, preds: Tensor, target: Tensor):
...
+
+        if self.subset_accuracy and not _check_subset_validity(self.mode):
+            self.subset_accuracy = False
```
Comment (Borda) [inline torchmetrics/classification/accuracy.py:232]: @arvindmuralie77 mind add test also here

File: torchmetrics/classification/accuracy.py [hunk 10]
```diff
@@ -141,18 +220,58 @@ def update(self, preds: Tensor, target: Tensor):
...
+
+        if self.subset_accuracy:
+            correct, total = _subset_accuracy_update(
```
Comment (Borda) [inline torchmetrics/classification/accuracy.py:235]: @arvindmuralie77 mind add test also here

File: torchmetrics/classification/accuracy.py [hunk 11]
```diff
@@ -141,18 +220,58 @@ def update(self, preds: Tensor, target: Tensor):
...
-        self.correct += correct
-        self.total += total
+            # Update states
+            if self.reduce != "samples" and self.mdmc_reduce != "samplewise":
+                self.tp += tp
+                self.fp += fp
+                self.tn += tn
+                self.fn += fn
+            else:
+                self.tp.append(tp)
```
Comment (Borda) [inline torchmetrics/classification/accuracy.py:261]: @arvindmuralie77 mind add test also here

File: torchmetrics/classification/accuracy.py [hunk 12]
```diff
@@ -141,18 +220,58 @@ def update(self, preds: Tensor, target: Tensor):
...
+                self.tn.append(tn)
+                self.fn.append(fn)
 
     def compute(self) -> Tensor:
         """
         Computes accuracy based on inputs passed in to ``update`` previously.
         """
-        return _accuracy_compute(self.correct, self.total)
+        if self.subset_accuracy:
+            return _subset_accuracy_compute(self.correct, self.total)
```
Comment (Borda) [inline torchmetrics/classification/accuracy.py:271]: @arvindmuralie77 mind add test also here

**Review by Borda — COMMENTED** [review]
a few more tests please and we are ready to land :]

File: tests/classification/test_accuracy.py [hunk 13]
```diff
@@ -200,3 +200,89 @@ def test_wrong_params(top_k, threshold):
...
+            tensor([2, 2, 1, 1]),
+            tensor([[1, 0, 1, 1], [0, 0, 1, 0]]),
+            tensor([[0.8, 0.2, 0.8, 0.7], [0.6, 0.4, 0.6, 0.5]])
+        )
+    ],
+)
+def test_different_modes(preds_mc, target_mc, preds_ml, target_ml):
+    acc = Accuracy()
+    acc(preds_mc, target_mc)
+    with pytest.raises(ValueError):
```
Comment (Borda) [inline tests/classification/test_accuracy.py:219]:
```suggestion
    with pytest.raises(ValueError, match="The `average` has to be one of"):
```
pls add some part of the error message so we are clear that it the one Error

File: torchmetrics/classification/accuracy.py (hunk 9 above)
Comment (arvindmuralie77) [reply torchmetrics/classification/accuracy.py:232]: I am confused as to what kind of test I need to add here? There is nothing to check here

File: torchmetrics/classification/accuracy.py (hunk 10 above)
Comment (arvindmuralie77) [reply torchmetrics/classification/accuracy.py:235]: Tests for subset_accuracy argument are already present.

File: torchmetrics/classification/accuracy.py (hunk 11 above)
Comment (arvindmuralie77) [reply torchmetrics/classification/accuracy.py:261]: Added some test cases for the cases where `reduce` equals `samples` and `mdmc_reduce` is not `sample_wise`.

File: torchmetrics/classification/accuracy.py (hunk 12 above)
Comment (arvindmuralie77) [reply torchmetrics/classification/accuracy.py:271]: I am not clear what kind of test cases need to be added here.

**Comment by arvindmuralie77** [comment]
I have added more test cases as mentioned

File: torchmetrics/classification/accuracy.py (hunk 12 above)
Comment (Borda) [reply torchmetrics/classification/accuracy.py:273]: is in any test `self.subset_accuracy == True`?

**Pull request merged**